File: leetcode/no0202_Happy_Number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
    let numString = n.toString();

    let total = 0;
    for (let i = 0; i < numString.length; i++) {
        let temp = Math.pow(parseInt(numString[i]), 2);
        console.log("+", temp)
        total += temp
    }
    console.log("Total= ", total, "from", numString);

    if (total == 2 || total == 4) {
        return false;
    } else if (total == 1) {
        return true;
    } else {
        return isHappy(total);
    }
};

File: kotlin/src/main/kotlin/no0202_Happy_Number.kt
fun main() {

    class Solution {
        fun isHappy(n: Int): Boolean {
            val numString = n.toString()
            var total = 0.0
            val size = numString.length
            for (i in 0 until size) {
                val char = numString[i]
                val temp = Math.pow(char.toString().toDouble(), 2.0)
                //println("#$i = $temp")
                total += temp
                //println("total = $total")
            }
            return when (total) {
                2.0, 4.0 -> {
                    false
                }
                1.0 -> {
                    true
                }
                0.0 -> {
                    false
                }
                else -> {
                    isHappy(total.toInt())
                }
            }
        }
    }

    val solution = Solution()

//    Input: n = 19
//    Output: true

    val result = solution.isHappy(0)
    println("OUTPUT: $result")

}
